### PR TITLE
auto-lock for trezor-core

### LINKED ISSUE #75: Configurable auto-lock feature
Trezor One have constant-time timeout for locking device, which may be too long for some and frustrating for heavy users who works in trusted environment.

Model T does not have auto-locking yet. Let's make the timeout configurable over ApplySettings() with the possibility to turn the feature off completely.

Also, the wake event should be a human interaction (touch event), not wire communication, to protect device from misbehaving app on desktop.

### PR DESCRIPTION
**Pull request by matejcik**
fixes #75 

this has WAY more commits than it rightly should.

The core functionality is f1c3172, which introduces a global idle timer. It is then possible to register an idle callback, which, after a configurable timeout (we already have autolock timeout fields in storage), engages the lock.

Another big part is a new `loop.spawn` syscall (docs included), which serves as a "controllable" subtask on the application level. All workflow tasks are now `spawn`ed, so that it is possible to shut them down when the autolock engages. (this also answers [this comment](https://github.com/trezor/trezor-firmware/pull/958#discussion_r425934798))

As a nicety, the lockscreen is now dimmed. I'm not sure about the backlight level, maybe we'll need to tweak that a little.

The remaining 80 % of the work is busywork:

- `workflow.kill_default()` was superseded by `workflow.close_others()`, because we now have the capability to shut down ALL tasks, not just the default.
- there is now a `button_request()` function that does just that in addition to `ButtonRequest/ButtonAck` calls. The logic here is that when you are calling `ButtonRequest`, you want to do something on the UI, and so you need to shut down whatever was there before you.
- `debuglink.wait_layout` has been reworked. If you are using `wait_layout`, you need to wait for *all* layout changes that happen. In exchange, it's resilient to race conditions: you _will_ receive all layout changes that happen.
- for this reason I needed to introduce `ButtonRequestType.PinEntry`, to cleanly distinguish the TT asking for PIN on screen.
- some UI test artifacts were cleared up by moving the "notify layout change" call to a more appropriate location

---

There's work to be done on the FIDO app, where another `loop.spawn` should be used for the keepalive loop, and that might simplify #729.

### CODE REVIEW

**Comment by matejcik** [comment]
open question: should FIDO use the global idle timer for its PIN timeout?

I think FIDO should respect the user-configured auto-lock delay, AS LONG AS it's 10 minutes (the default) or less. Opinions?

If it respects the autolock fully, we can remove the custom timeout feature of the PIN dialog.

We can also replace the custom timeout by an idle callback, for a *slightly different* result. As implemented, the PIN timeout applies to FIDO unlocks only. With an idle callback, it would apply to all device activity.
Do we want that?

**Comment by tsusanka** [comment]
UI changes:

https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/571346102/artifacts/master_diff/index.html

**Comment by matejcik** [comment]
the UI report seems wrong -- as if tests copied a lot of their screens from their longer friends

**Comment by matejcik** [comment]
fixed UI report: https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/572864736/artifacts/master_diff/index.html

File: core/src/apps/common/__init__.py [hunk 1]
```diff
@@ -1,3 +1,19 @@
...
+    from trezor.messages.ButtonRequest import EnumTypeButtonRequestType
+
 HARDENED = const(0x80000000)
+
+
+async def button_request(
+    ctx: wire.GenericContext, code: EnumTypeButtonRequestType = ButtonRequestType.Other
+) -> None:
+    workflow.close_others()
+    await ctx.call(ButtonRequest(code=code), ButtonAck)
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/common/__init__.py:19]: Correct. Sending ButtonRequest when you're not drawing anything is incorrect.

File: core/src/apps/common/passphrase.py [hunk 2]
```diff
@@ -26,6 +26,7 @@ async def get(ctx: wire.Context) -> str:
 
 
 async def _request_from_user(ctx: wire.Context) -> str:
+    workflow.close_others()  # request exclusive UI access
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/common/passphrase.py:29]: It's because of `_entry_dialog()`. We are now in the passphrase flow and that always puts _something_ on screen

File: core/src/trezor/loop.py [hunk 3]
```diff
@@ -47,14 +47,22 @@
...
     """
     Schedule task to be executed with `value` on given `deadline` (in
     microseconds).  Does not start the event loop itself, see `run`.
     Usually done in very low-level cases, see `race` for more user-friendly
     and correct concept.
+
+    If `reschedule` is set, updates an existing entry.
     """
+    if reschedule:
+        _queue.discard(task)
```
Comment (andrewkozlik) [inline core/src/trezor/loop.py:75]: I haven't given this much thought, but I wonder, would it make sense to schedule one task multiple times? What happens if we do? Maybe schedule() should search the queue for the task, if it finds that it's already scheduled, then reschedule it. In other words maybe there is no need for the reschedule flag.

File: core/src/apps/management/apply_settings.py [hunk 4]
```diff
@@ -36,17 +41,27 @@ async def apply_settings(ctx, msg):
     if msg.display_rotation is not None:
         await require_confirm_change_display_rotation(ctx, msg.display_rotation)
 
+    if msg.auto_lock_delay_ms is not None:
+        msg.auto_lock_delay_ms = max(
+            msg.auto_lock_delay_ms, storage.device.AUTOLOCK_DELAY_MINIMUM
+        )
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/management/apply_settings.py:47]: This copies T1 behavior. We could change it, but we'd have to change T1 too, probably.

File: core/src/trezor/loop.py (hunk 3 above)
Comment (matejcik) [reply core/src/trezor/loop.py:75]: `discard()` is a linear search, and `schedule()` is called relatively often.
I'm not sure how expensive this actually is on real hardware, but I was working off the assumption that this is the place for micro-optimizations ;)

Comment (matejcik) [reply core/src/trezor/loop.py:75]: Re multiple scheduled tasks, i would consider that a new feature/requirement.
Currently we don't need it, and the queue would indeed be more correct if it explicitly disallowed that. There are some interesting usecases that could make use of a task scheduled multiple times, but then we have a problem with maintaining the state of the task object.

The main reason the queue does not reject that, AFAICT, is performance.

Comment (matejcik) [reply core/src/trezor/loop.py:75]: Re "what happens": tasks are running coroutines/generators. You can safely schedule a task multiple times if, e.g., there is an infinite loop running inside it.
(a finished generator will simply die off, this is handled properly and does not cause a crash)
A complication is that the syscall machinery makes sure that the task is rescheduled whenever it awaits something. So e.g., if you schedule a task twice, it will run twice as often - and more importantly, the return values from the awaited syscalls will get intermingled.

In short, "bad things" is what would happen.
For that reason `loop.schedule` should generally not be used by application code.

File: core/src/apps/homescreen/lockscreen.py [hunk 5]
```diff
@@ -33,6 +39,33 @@ def render_lock(self) -> None:
...
+        # repaint (we can be running the same layout instance multiple times)
+        # and paint it.
+        ui.display.clear()
+        self.on_render()
+
+        if __debug__ and self.should_notify_layout_change:
+            from apps.debug import notify_layout_change
+
+            self.should_notify_layout_change = False
+            notify_layout_change(self)
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/homescreen/lockscreen.py:60]: That doesn't really seem worth it? This is a one-off. If we end up doing something like this in more places, the requirements will become clearer.

(fwiw if we want to get rid of this, we could change `backlight_fade(style.BACKLIGHT_NORMAL)` to `backlight.fade(self.backlight_level)` and remove the copy-paste. It didn't seem worth polluting all Layouts with a new variable, but maybe the cost is negligible.
same thing with the `loop.sleep()` which could also be a class or instance variable, and in case of homescreen and lockscreen could be changed to something like `loop.DO_NOT_RESCHEDULE`)

File: core/src/trezor/workflow.py [hunk 6]
```diff
@@ -130,3 +135,70 @@ def _finalize_default(task: loop.Task, value: Any) -> None:
...
+
+        `idle_timer.set()` also counts as an activity, so all running idle timers are
+        reset.
+        """
+        if callback in self.tasks:
+            loop.close(self.tasks[callback])
+
+        self.timeouts[callback] = timeout_ms * 1000
+        self.tasks[callback] = self._timeout_task(callback)
+        self.touch()
```
Comment (andrewkozlik) [inline core/src/trezor/workflow.py:213]: TBH I don't understand what's going on here, but why should set() call self.touch()? I thought touch() resets the idle timer. It just seems surprising that adding a callback would do that.

File: core/src/storage/device.py [hunk 7]
```diff
@@ -38,6 +38,8 @@
 # fmt: on
 
 HOMESCREEN_MAXSIZE = 16384
+AUTOLOCK_DELAY_MINIMUM = 10 * 1000  # 10 seconds
+AUTOLOCK_DELAY_DEFAULT = 10 * 60 * 1000  # 10 minutes
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/storage/device.py:42]: yes

**Comment by matejcik** [comment]
> Btw this is not turned on on default for both old and new users correct? I should talk to Product whether we would like to turn this on in Suite by default (aka during onboarding).

Very good point.

With this PR, auto-lock is _always enabled_. If the user never configured it, the default timeout of 10 minutes is used.
IMHO this is what we actually want, the "always unlocked" behavior from previous revisions was not desirable. But it is a Product consideration.

File: core/src/trezor/workflow.py (hunk 6 above)
Comment (matejcik) [reply core/src/trezor/workflow.py:213]: It resolves an ambiguity wrt idleness and relative ordering of the idle tasks.

Namely: if `self.touch()` didn't happen, should this schedule the task to 10 minutes from _now_, or from _last touch_? If from _now_, other tasks scheduled to 10 minutes might happen at different time (but only for the first time, next time they will be the same). If from _last touch_, and more than 10 minutes elapsed, should the task run right now, or only next time an activity is registered and 10 minutes elapse?

I mean, all of this is pointless because adding a task only happens _during_ an activity. But counting `set()` as an activity seems very clean.
We can easily change this if there are new requirements.

Comment (matejcik) [reply core/src/trezor/workflow.py:213]: i'll add a comment to the code about this

Comment (matejcik) [reply core/src/trezor/workflow.py:219]: added in 20d7fe0

File: core/src/apps/common/passphrase.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/common/passphrase.py:29]: I mean why it is not in `_entry_dialog` then. In `_entry_dialog` for `_request_on_host` and `_request_on_device` has it in `button_request`.

File: core/src/apps/homescreen/lockscreen.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/homescreen/lockscreen.py:60]: > backlight.fade(self.backlight_level)

I would do it. My concern is that if we change something in `Layout.handle_rendering` we forget to add it here. IMHO.

File: core/src/apps/management/apply_settings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/apply_settings.py:47]: Ok let's not worry about that for now then.

File: core/src/trezor/workflow.py [hunk 8]
```diff
@@ -26,23 +26,23 @@
 default_constructor = None  # type: Optional[Callable[[], loop.Task]]
 
 
-def on_start(workflow: loop.Task) -> None:
+def _on_start(workflow: loop.spawn) -> None:
     """
-    Call after creating a workflow task, but before running it.  You should
+    Called after creating a workflow task, but before running it.  You should
     make sure to always call `on_close` when the task is finished.
```
Comment (andrewkozlik) [inline core/src/trezor/workflow.py:32]: Seems to me this comment is no longer relevant, since calling `_on_close()` is automated.

Comment (andrewkozlik) [reply core/src/trezor/workflow.py:32]: Although there is an `on_close()` method (without the initial underscore) elsewhere, but as far as I can tell, that's not what is meant here.

File: core/src/trezor/loop.py [hunk 9]
```diff
@@ -421,3 +443,118 @@ def _schedule_take(self, taker: Task) -> None:
...
+
+        If another caller is awaiting its result it will get a TaskClosed exception.
+        If the task was already finished, the call has no effect.
+        """
+        if not self.finished:
+            if __debug__:
+                log.debug(__name__, "close spawned task: %s", self.task)
+            close(self.task)
+
+    def on_close(self, finalizer_callback: Callable[["spawn"], None]) -> None:
```
Comment (andrewkozlik) [inline core/src/trezor/loop.py:544]: The function name here is confusing. This function is not called upon close, it is used to set the callback which should be called upon close. How about `set_finalizer()`?

**Comment by tsusanka** [comment]
> Unable to test. When I load this firmware on a physical device I get a black screen after startup.

That's because this PR is missing b01b24f09035791519d747cb640cd0e13292c8da. That is why force-pushes are needed sometimes :/. Maybe let's finish the review and rebase then?

File: core/src/apps/common/passphrase.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/common/passphrase.py:29]: It makes more sense to me like this.

Before (no "request UI exclusivity" exists): we draw things as needed, and as a special case, in `_entry_dialog` we _also_ need to make sure that default is killed.

Now: in `_request_from_user` we are entering the UI-enabled part of passphrase entry. Request UI exclusivity because we need it in either case: the fact that `button_request()` does the same thing later becomes an implementation detail.

File: core/src/apps/homescreen/lockscreen.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/homescreen/lockscreen.py]: done in 7a2b8e748109165d649d3a01f71362519172de20

File: core/src/trezor/workflow.py (hunk 8 above)
Comment (matejcik) [reply core/src/trezor/workflow.py]: done in 62552006a2cacbd677bcf341163566da7544c03f

File: core/src/trezor/loop.py (hunk 9 above)
Comment (matejcik) [reply core/src/trezor/loop.py]: done (in a series of fixups)

**Comment by matejcik** [comment]
I tried to rebase on master and run this on a real hardware. Currently the device freezes after the initial startup, I'm investigating.

**Comment by matejcik** [comment]
i fixed the scheduler issue by lowering resolution, and cherry-picked b01b24f so that this branch can now run on real hardware. Ready for review again.

**Comment by andrewkozlik** [comment]
1. I don't think that the screen should be dimmed in the "Not connected" state. I am OK with it dimming after a few minutes of inactivity, but not from the start.
2. Is it just my impression, or has the "Not connected"/"Locked" label shifted downwards? It doesn't look centred to me (in relation to the standard backgrounds we have).

**Comment by matejcik** [comment]
> 2. Is it just my impression, or has the "Not connected"/"Locked" label shifted downwards? It doesn't look centred to me (in relation to the standard backgrounds we have).

I don't think I moved this. Compared to 2.3.0 it looks the same.

**Comment by andrewkozlik** [comment]
> > 1. Is it just my impression, or has the "Not connected"/"Locked" label shifted downwards? It doesn't look centred to me (in relation to the standard backgrounds we have).
> 
> I don't think I moved this. Compared to 2.3.0 it looks the same.

I see it's also not showing the label, so that might be causing a shift in the background image.

**Master**
master
**Autolock**
autolock-connect autolock-unlock

**Comment by matejcik** [comment]
Correct. I didn't move the texts, but I did move the homescreen image up, because I'm using the same code as the homescreen (which has a label on bottom).

I'm fixing this as we speak.

**Comment by matejcik** [comment]
rendering issues in homescreens (moved image; repainting on touch) and backlight on initial "not connected" are fixed in 5910c53

File: core/src/storage/device.py [hunk 10]
```diff
@@ -40,6 +40,8 @@
 HOMESCREEN_MAXSIZE = 16384
 AUTOLOCK_DELAY_MINIMUM = 10 * 1000  # 10 seconds
 AUTOLOCK_DELAY_DEFAULT = 10 * 60 * 1000  # 10 minutes
+# autolock intervals larger than AUTOLOCK_DELAY_MAX cause issues in the scheduler
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/storage/device.py]: done in a00abbf152aef1b46556a9702b941be7410a7c70

File: core/src/trezor/loop.py [hunk 11]
```diff
@@ -202,23 +222,26 @@ def handle(self, task: Task) -> None:
...
+
+
 class sleep(Syscall):
-    """
-    Pause current task and resume it after given delay.  Although the delay is
-    given in microseconds, sub-millisecond precision is not guaranteed.  Result
-    value is the calculated deadline.
+    """Pause current task and resume it after given delay.
+
+    Resultvalue is the calculated deadline.
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/trezor/loop.py]: fixed (i think?) in 50052a68eac8506fdde30447b7a02bf0ffc3948b
